components/react/hooks/src/useRequest/index.js

```diff
@@ -0,0 +1,179 @@
+import {useCallback, useEffect, useReducer, useRef} from 'react'
+
+export const STATUSES = {
+  IDLE: 'idle',
+  LOADING: 'loading',
+  ERROR: 'error',
+  SUCCESS: 'success'
+}
+
+const ACTIONS = {
+  FETCHING: 'fetching',
+  LOADING: 'loading',
+  ERROR: 'error',
+  SUCCESS: 'success',
+  REPLACE: 'replace',
+  RESET: 'reset'
+}
+
+const initialState = {
+  status: STATUSES.LOADING,
+  data: null,
+  error: null,
+  isFetching: false
+}
+
+const getInitialState = ({initialData, initialStatus}) => {
+  const data = typeof initialData === 'function' ? initialData() : initialData
+  const status = data !== undefined ? STATUSES.SUCCESS : initialStatus || initialState.status
+
+  return {
+    status,
+    data,
+    error: null
+  }
+}
+
+const reducer = (state, action) => {
+  switch (action.type) {
+    case ACTIONS.FETCHING:
+      return {
+        ...state,
+        isFetching: true
+      }
+    case ACTIONS.LOADING:
+      return {
+        ...initialState,
+        isFetching: true,
+        data: state.data
+      }
+    case ACTIONS.ERROR:
+      return {
+        ...initialState,
+        status: STATUSES.ERROR,
+        error: action.payload.error,
+        data: state.data
+      }
+    case ACTIONS.SUCCESS:
+      return {
+        ...initialState,
+        status: STATUSES.SUCCESS,
+        data: action.payload.data
+      }
+    case ACTIONS.REPLACE:
+      return {
+        ...initialState,
+        status: STATUSES.SUCCESS,
+        data: action.payload.isHard ? action.payload.data : {...state.data, ...action.payload.data}
+      }
+    case ACTIONS.RESET:
+      return {
+        ...initialState,
+        status: action.payload.initialStatus
+      }
+    default:
+      return state
+  }
+}
+
+const useRequest = (request, {initialStatus, initialData, onSuccess, onError} = {}) => {
+  const [{status, data, ...others}, dispatch] = useReducer(reducer, getInitialState({initialStatus, initialData}))
+  const requestRef = useRef(request)
+  const onSuccessRef = useRef(onSuccess)
+  const onErrorRef = useRef(onError)
```

Comment:
Why do you need to set these props into a ref?

Reply:
I see, I guess is to avoid the dependency in the `useCallback` for the `fetch` function so it does not get recomputed if the callbacks reference change. But what happens with this internal onSuccessRef if the onSuccess reference changes? Imagine a case where you need to recreate a new onSuccess callback when, for example, the `id` parameter changes so your callback looks like this: `() => console.log(id)` and the `id` comes from the outer scope. Does the onSuccessRef get updated with the new value?